### PR TITLE
New IPAAnsibleModule class

### PR DESCRIPTION
**Pull request by t-woerner**
ansible_freeipa_module: New IPAAnsibleModule class
    
    This class is an extended version of the Ansible Module that provides
    IPA specific methods to simplify module generation.
    
    Simple example:
    
        from ansible.module_utils.ansible_freeipa_module import \
            IPAAnsibleModule
    
        def main():
            ansible_module = IPAAnsibleModule(
                argument_spec=dict(
                      name=dict(type="str", aliases=["cn"], default=None),
                      state=dict(type="str", default="present",
                                 choices=["present", "absent"]),
                ),
            )
    
            # Get parameters
            name = ansible_module.params_get("name")
            state = ansible_module.params_get("state")
    
            # Connect to IPA API
            with ansible_module.ipa_connect():
    
                # Execute command
                if state == "present":
                    ansible_module.ipa_command(["command_add", name, {}])
                else:
                    ansible_module.ipa_command(["command_del", name, {}])
    
            # Done
    
            ansible_module.exit_json(changed=True)
    
        if __name__ == "__main__":
            main()

ipamodule[+member].py.in: Use IPAAnsibleModule class, reduce calls
    
    Use IPAAnsibleModule and ipamodule_base_docs in the templates of
    utils/new_module.
    
    ipaadmin_password lines in the examples have been added, ipaadmin_
    variables are handled by IPAAnsibleModule, ansible_module.params_get is
    used to get the parameters and ansible_module.ipa_connect is used to
    simplify the module.
    
    ipamodule+member.py.in is additionally using gen_add_list and
    gen_intersection_list to reduce the command calls to the changes only.

*modules*: Use IPAAnsibleModule class
    
    ipaadmin_variables are handled by IPAAnsibleModule,
    ansible_module.params_get is used to get the parameters and
    ansible_module.ipa_connect is used to simplify the module.

### CODE REVIEW

**Comment by rjeffman** [comment]
This is a very nice patch that removes 600+ lines of code! ;-)

I have a few suggestions:
* The IPAModuleClass should be declared before FreeIPABaseModule, so it can be used as a base class for the second one;
* The methods could use some documentation, at least, for the parameters;
* The methods not using `self` could be declared `@staticmethod`.

I [proposed a PR](https://github.com/t-woerner/ansible-freeipa/pull/1) with this changes, as they also allow other modules, currently based on FreeIPABaseModule to use the same infrastructure and document fragment defined in this PR.

**Comment by t-woerner** [comment]
> I [proposed a PR](https://github.com/t-woerner/ansible-freeipa/pull/1) with this changes, as they also allow other modules, currently based on FreeIPABaseModule to use the same infrastructure and document fragment defined in this PR.
The PR is merged, thank you.

File: plugins/modules/iparole.py [hunk 1]
```diff
@@ -285,8 +264,9 @@ def ensure_role_with_members_is_present(module, name, res_find, action):
             del_members[key] = [to_text(item) for item in del_list]
 
     service = [
-        to_text(svc) if '@' in svc else ('%s@%s' % (svc, api_get_realm()))
-        for svc in (module_params_get(module, 'service') or [])
+        to_text(svc) if '@' in svc else ('%s@%s' % (svc,
+                                                    module.ipa_get_realm()))
+        for svc in (module.params_get('service') or [])
```
Comment (rjeffman) [inline plugins/modules/iparole.py:269]: I like the format suggestion by `black` for this code snippet:
```python
    to_text(svc)
    if "@" in svc
    else ("%s@%s" % (svc, module.ipa_get_realm()))
    for svc in (module.params_get("service") or [])
```
I find it easier to read.

**Review by rjeffman — COMMENTED** [review]
Function `execute_api_command` (plugins/module_utils/ansible_freeipa_module.py:257) is not used anymore and should be deprecated (or removed at all).

Despite my comments on this review, this change is very welcome!

File: plugins/module_utils/ansible_freeipa_module.py [hunk 2]
```diff
@@ -568,7 +568,258 @@ def names(self):
...
+            try:
+                if not valid_creds(self, self.ipaadmin_principal):
+                    ccache_dir, ccache_name = temp_kinit(
+                        self.ipaadmin_principal, self.ipaadmin_password)
+                api_connect(context)
+            except Exception as e:
+                self.fail_json(msg=str(e))
+            else:
+                try:
+                    yield ccache_name
```
[the remark this answers is not in the document]
Comment (t-woerner) [reply plugins/module_utils/ansible_freeipa_module.py:655]: ipavault needs to set `os.environ["KRB5CCNAME"]` using ccache_name.

File: plugins/module_utils/ansible_freeipa_module.py [hunk 3]
```diff
@@ -568,7 +568,258 @@ def names(self):
...
+            # pylint: disable=super-with-arguments
+            super(IPAAnsibleModule, self).__init__(*args, **kwargs)
+
+            # ipaadmin vars
+            self.ipaadmin_principal = self.params_get("ipaadmin_principal")
+            self.ipaadmin_password = self.params_get("ipaadmin_password")
+
+            # Attributes to store kerberos credentials (if needed)
+            self.ccache_dir = None
+            self.ccache_name = None
```
[the remark this answers is not in the document]
Comment (t-woerner) [reply plugins/module_utils/ansible_freeipa_module.py:630]: Ack, will be removed.

File: plugins/module_utils/ansible_freeipa_module.py [hunk 4]
```diff
@@ -568,7 +568,258 @@ def names(self):
...
+                _spec = kwargs["argument_spec"]
+                _spec.update(ipamodule_base_spec)
+                kwargs["argument_spec"] = _spec
+
+            # pylint: disable=super-with-arguments
+            super(IPAAnsibleModule, self).__init__(*args, **kwargs)
+
+            # ipaadmin vars
+            self.ipaadmin_principal = self.params_get("ipaadmin_principal")
+            self.ipaadmin_password = self.params_get("ipaadmin_password")
```
[the remark this answers is not in the document]
Comment (t-woerner) [reply plugins/module_utils/ansible_freeipa_module.py:626]: Ack, will be changed.

File: plugins/modules/iparole.py (hunk 1 above)
Comment (t-woerner) [reply plugins/modules/iparole.py:269]: What do you think about using `afm_convert` for all methods in IPAAnsibleModule. This way we could get rid of all `to_text` calls in the modules.

File: plugins/module_utils/ansible_freeipa_module.py [hunk 5]
```diff
@@ -568,7 +568,258 @@ def names(self):
...
+            # Done
+
+            ansible_module.exit_json(changed=True)
+
+        if __name__ == "__main__":
+            main()
+
+        """
+
+        def __init__(self, *args, **kwargs):
```
[the remark this answers is not in the document]
Comment (rjeffman) [reply plugins/module_utils/ansible_freeipa_module.py:581]: I agree.

File: plugins/modules/iparole.py (hunk 1 above)
Comment (rjeffman) [reply plugins/modules/iparole.py]: No having to use `to_text` on the modules would be nice. Sometimes this errors show up very late.

Comment (t-woerner) [reply plugins/modules/iparole.py]: Let's test the integration of _afm_convert and to_text in another PR. It might take some time to get all sorted out.

**Review by rjeffman — APPROVED** [review]
lgtm.

**Pull request merged**